### DISCUSSION TITLE
HCL Colorspace Visualization

### DISCUSSION THREAD

**Discussion by eitanlees** [colour-science/colour#1096] [Q&A]
Hello, 
I am new to colour and learning about different colorspaces. I am a little confused and wanted to ask a few beginner questions. 

I was trying to vary different parameters of a color space between [0, 1] and create a visualization of the corresponding output. 

```python
import colour
import numpy as np
import matplotlib.pyplot as plt

# colorspace = 'RGB'
colorspace = 'HSV'
# colorspace = 'HSL'
# colorspace = 'HCL'

grid = np.mgrid[0:1:100j,0:1:100j, 0:1:100j]

fig, axs = plt.subplots(2, 4, figsize=(10, 5), constrained_layout=True)

for i, ax in enumerate(axs.ravel()):
    
    with colour.utilities.suppress_warnings(colour_usage_warnings=True):
        RGB = colour.convert(grid[:, i*10, :, :].T, colorspace, 'RGB')
        
    ax.imshow(RGB)
    ax.set_axis_off()
    ax.invert_yaxis()
    ax.set_aspect('equal')
```
<img width="1007" alt="Screenshot 2023-01-20 at 17 07 09" src="https://user-images.githubusercontent.com/2041969/213813728-4c72ba71-5c9b-4f07-80d2-a7e09da7bdaa.png">

This worked well for RGB, HSV, and HSL but did not work for HCL. I am not sure what is going on. Reading the documentation for the [`HCL_to_RGB`](https://colour.readthedocs.io/en/latest/generated/colour.HCL_to_RGB.html) mentions a domain and range of [0,1] for both color spaces, so I was confused by this behavior. 

<img width="668" alt="HCL-docs" src="https://user-images.githubusercontent.com/2041969/213813827-4b8e9dd4-4402-4865-9d67-c5c0b8bd2dec.png">

Also in the example they are using a number outside of this range. I am not sure I understand what is the correct range of values the HCL color space expects. 

I am interested in creating a visualization similar to this one I've found online 
<img width="1228" alt="HCL-plot" src="https://user-images.githubusercontent.com/2041969/213814206-cf0e32b7-f680-4971-8ceb-e83ba1af4db3.png">
([source](https://clauswilke.com/art/post/colors-color-spaces))

Any advice on how to create such a plot is appreciated 🙂 

Thanks!

**Comment by thomasmansencal** [comment]
Hi @eitanlees,

This one is an oddity and I have been wondering if we should not get rid of it:

There are two references for it:
  1. Sarifuddin, M., & Missaoui, R. (2005a). A New Perceptually Uniform Color Space with Associated Color Similarity Measure for ContentBased Image and Video Retrieval.
  2. Sarifuddin, M., & Missaoui, R. (2005b). HCL: a new Color Space for a more Effective Content-based Image Retrieval. http://w3.uqo.ca/missaoui/Publications/TRColorSpace.zip

We use the second one as it provides an inverse but there are some fixes / changes compared to 1. in the forward model.

If you use some RGB values, you will see that hue is in range [-1/3 PI, PI] and Lightness in range ~[0, 135/255]:

```python
P = [
    [0, 0, 0],
    [1, 0, 0],
    [0, 1, 0],
    [0, 0, 1],
    [1, 1, 0],
    [0, 1, 1],
    [1, 0, 1],
    [1, 1, 1],
]

colour.convert(P, "RGB", "HCL")
```

```python
array([[ 0.        ,  0.        ,  0.        ],
       [ 0.        ,  0.66666667,  0.5       ],
       [ 2.0943951 ,  0.66666667,  0.5       ],
       [ 0.        ,  0.66666667,  0.5       ],
       [ 0.        ,  0.66666667,  0.5       ],
       [ 3.14159265,  0.66666667,  0.5       ],
       [-1.04719755,  0.66666667,  0.5       ],
       [ 0.        ,  0.        ,  0.53045453]])
```

As for the reason the domain / range is given as being [0, 1], it is something we should maybe update here but it is worth being aware of this: https://colour.readthedocs.io/en/develop/basics.html#domain-range-scales, i.e. they are _soft_ domain / ranges.

Playing quickly with the model and comparing the two references, one notable change is the computation of `L` with  L = (Q * Max + (1 - Q) * Min) / 2 for **1.** `L = (Q * Max + (Q - 1) * Min) / 2` for **2.**. The former pegs `L` to 0.5 instead of 0.53045453 for `RGB = [1, 1, 1]` and does not roundtrip at all with the inverse while the latter does a bit better and matches the 0.53045453 * 255 ~= 135 output that the paper mentions!
